### PR TITLE
SLAM_cone_detection (with description)

### PR DESCRIPTION
**Pull request by YuriKun1033**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimisation

## Description

!UNTESTED!
using SLAM gridmap to detect cones (2D)
use of LiDAR DBSCAN point clustering process

This script takes the SLAM gridmap that gets created, recognizes and clusters points using DBSCAN
the clustered points then get recognised as cones if they are within certain parameters
the gridmap get a new origin point with each update the map gets

## Documentation

https://www.notion.so/qut-motorsports/SLAM-Cone-Detection-9b6e0e06287b4d12b28a27e52a949ccd

### Usability concerns or breaking changes?

None

###  Screenshots

image
(cone overlay onto SLAM toolbox gridmap)

### CODE REVIEW

File: src/navigation/slam_gridmap/package.xml
Comment (b1n-ch1kn) [inline src/navigation/slam_gridmap/package.xml:1]: General comment on filling out package info - it is useful for traceability to see who worked on what code.
Filling out description, maintainer (use team email), license etc.
Check other examples of our pkgs for that.

File: src/navigation/slam_gridmap/package.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="claudiakiss005@gmail.com">yurikun1033</maintainer>
+  <license>TODO: License declaration</license>
+
+  <test_depend>ament_copyright</test_depend>
+  <test_depend>ament_flake8</test_depend>
+  <test_depend>ament_pep257</test_depend>
+  <test_depend>python3-pytest</test_depend>
+
```
Comment (b1n-ch1kn) [inline src/navigation/slam_gridmap/package.xml:14]: Probably can ditch the tests and the test folder (they're created automatically by ros)

File: src/navigation/slam_gridmap/package.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>slam_gridmap</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="claudiakiss005@gmail.com">yurikun1033</maintainer>
+  <license>TODO: License declaration</license>
+
```
Comment (b1n-ch1kn) [inline src/navigation/slam_gridmap/package.xml:9]: Add dependencies for the msgs we're using - see other examples of our package's XMLs for how to do this

File: src/navigation/slam_gridmap/setup.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    data_files=[
+        ("share/ament_index/resource_index/packages", ["resource/" + package_name]),
+        ("share/" + package_name, ["package.xml"]),
+    ],
+    install_requires=["setuptools"],
+    zip_safe=True,
+    maintainer="yurikun1033",
+    maintainer_email="claudiakiss005@gmail.com",
+    description="TODO: Package description",
+    license="TODO: License declaration",
```
Comment (b1n-ch1kn) [inline src/navigation/slam_gridmap/setup.py:18]: This info is copy paste from the XML fields of the same names

File: src/perception/lidar_pipeline/lidar_pipeline/node_lidar_detector.py [hunk 4]
```diff
@@ -173,7 +173,7 @@ def callback(self, msg: PointCloud2):
         # Step 1
 
         # Convert PointCloud2 message from LiDAR sensor to numpy array
-        dtype_list: list = fields_to_dtype(msg.fields, msg.point_step)  # x, y, z, intensity, ring
+        dtype_list: list = fields_to_dtype(msg.fields, msg.point_step)  # x, y, z, intensity, ringr
```
Comment (b1n-ch1kn) [inline src/perception/lidar_pipeline/lidar_pipeline/node_lidar_detector.py:176]: Not critical, in the future, double check what files and lines are being staged for the commit and only select stage file on only the ones the feature is about.
(ie. not this lidar file)

**Review by b1n-ch1kn — APPROVED** [review]

**Review by AlexCB192 — APPROVED** [review]
Request looks good. Any updates to testing methodology for next track log them in the respective thread

**Pull request merged**